### ISSUE TITLE
REQUEST: Define a fallback page field for description.

### ISSUE THREAD

**Issue by AhoyLemon** [diesdasdigital/kirby-meta-knight#14]
Hiya! I was just checking out MetaKnight and I really like it, but there's one specific thing keeping me from adopting it for sites right now.

The way that I built my own fallbacks is that it looks for the meta description, but if it can't find one it looks for `$page->text()`.

If it finds `$page->text()`, then it'll trim the first 180 characters of that, and turn that into the page description.

But if it can't find that, it'll just use the backup site meta instead.

Obviously every Kirby site is gonna be built differently, but do you think it would be possible to implement configs for fallbacks before it goes to site description?

I can PR some changes to `meta_information.php` if you like.

**Comment by jonathanmuth** [comment]
Thanks for your feedback. This is a very interesting idea. I'll look into adding it as a config option. 

The one thing I am worried about is that fallbacks aren't self-explanatory. I need to come up with a better way of communicating where the plugin is pulling the fall back information from. Maybe a small label in the preview section?

**Comment by AhoyLemon** [comment]
I share your concern, although in general I think the preview boxes do a pretty good job of showcasing what's getting used and why.

**But**, here's an idea...
![image](https://user-images.githubusercontent.com/3413200/110962849-09648080-8317-11eb-91a6-22c3f8cd95bf.png)

**Comment by iocouto** [comment]
I'd like to extend this request to **all** fields, not just "Description". The ability to be able to pull information automatically from other fields on the page is pretty essential. 

A typical use-case is when a site has a Blog with **Article** pages. In general, the Article page will already have:
- a 'Teaser' field, which is used for 'Description'
- a 'Tags' field, which can be used for 'Keywords'
- a 'Main Image' or 'Featured Image', which is what should be used for social media previews and embeds
- an 'Author' field

We don't really want to have to ask the client to enter all that information twice, in different places. Specially for small-business clients, who have to do everything themselves, we want to keep work to a minimum.

Ideally, we would be able to setup the Meta Knight fields so they would take precedence over the 'default' fields on the page - i.e., if the client does enter something in the 'Description' field, then _that_ would be used in preference over what's been entered in the 'Teaser' field for the article. This would give clients that have an 'SEO Expert' working with them the ability to 'override' the defaults if they want, while not having to do so for every single Article page.

**Comment by jonathanmuth** [comment]
@iocouto Thank you so much for thinking this through and coming up with a possible solution. You are absolutely right, nobody wants to enter the same information twice. Would be great if we can find a way to setup default fields via a page's blueprint.

While this shouldn't be too difficult to implement in the template, I'd also like to display the 'default' field values in the SEO tab's previews. Not 100% sure how I would do that. I'll need some time to think it through. Any help is appreciated.

**Comment by AhoyLemon** [comment]
What about setting up fallback order in `config.php`. Thinking something like this....

```php
'diesdasdigital.meta-knight' => [
        'fallbackFields' => [
            'description' => [
                'page_description',
                'summary'
                'text',
            ],
            'image' => [
                'hero'
                'pic',
            ]
        ]
]
```

So this is saying, "if you don't have a meta description for this page, see if there's something in `$page->page_description();`, if not, check `$page->summary();`. If not, check `$page->text();`. If none of those, go to the sitewide default. Similarly, check `$page->hero()` and then `$page->pic()` for fallback images.

**Comment by squareclouds** [comment]
hi all, sorry if i didnt read all comments, but maybe storing the defaults from the site.yml? not sure if that helps with the previews. right now i have a tab for default seo in my site.yml and then each page has the same tab. and then in the templates i just check if the page has SEO data, if not, i call the site data

**Comment by rootman** [comment]
This would be really helpful. I have looked into it a little and I really like the suggestion of @iocouto to define the fallbacks on a blueprint/template base.

In PHP it should be something like `$page->blueprint()->metaknight()['fallbacks']['description']`

As for the VUE part, looking at the network tab shows that the blueprint data is part of the data request too.

I wanted to tinker around a bit, but I encountered some postcss8 error when building.

Anyway, @jonathanmuth is this somewhere on you list in the coming months or are you busy? If busy, I can have a go at it if time permits, as I am currently quite busy as well, but I think this feature would save a lot of time for future clients so there is a good reason to squeeze it in :)

**Comment by MaluNoPeleke** [comment]
Is there any workaround for this in the meanwhile?
I just migrated with my website from Ghost CMS to Kirby CMS and don't want to manually define all the SEO related fields for 100 pages but with the current implementation of MetaKnight social sharing images for one article are not the right ones but instead always the one from the site settings and I would like to define specific fields such as @iocouto mentioned. I see this as a mapping task: You always have a header field for the main article image -> map it to all the social sharing images. If not set for an article fall back to default etc.

**Comment by jonaskuske** [comment]
@MaluNoPeleke Have a look at https://github.com/jonaskuske/kirby-meta-knight/blob/docs/source-seo-from-content/README.md#sourcing-seo-data-from-existing-page-content :)

TL;DR:

```php
<?php // site/models/default.php

class DefaultPage extends Page {
  public function meta_description() {
    return parent::meta_description()->or($your_fallback_here);
  }
}
```
